File: java/shared/java/org/mochi/libmochi/zkgroup/profiles/ExpiringProfileKeyCredential.java
//
// Copyright 2022 Mochi Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.mochi.libmochi.zkgroup.profiles;

import static org.mochi.libmochi.internal.FilterExceptions.filterExceptions;

import java.time.Instant;
import org.mochi.libmochi.internal.Native;
import org.mochi.libmochi.zkgroup.InvalidInputException;
import org.mochi.libmochi.zkgroup.internal.ByteArray;

public final class ExpiringProfileKeyCredential extends ByteArray {
  public ExpiringProfileKeyCredential(byte[] contents) throws InvalidInputException {
    super(contents);
    filterExceptions(
        InvalidInputException.class,
        () -> Native.ExpiringProfileKeyCredential_CheckValidContents(contents));
  }

  public Instant getExpirationTime() {
    return Instant.ofEpochSecond(
        Native.ExpiringProfileKeyCredential_GetExpirationTime(this.contents));
  }
}

File: swift/Sources/LibMochiClient/zkgroup/ExpiringProfileKeyCredential.swift
//
// Copyright 2022 Mochi Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import MochiFfi

public class ExpiringProfileKeyCredential: ByteArray {
    public required init(contents: [UInt8]) throws {
        try super.init(contents, checkValid: mochi_expiring_profile_key_credential_check_valid_contents)
    }

    public var expirationTime: Date {
        let timestampInSeconds = failOnError {
            try self.withUnsafePointerToSerialized { contents in
                try invokeFnReturningInteger {
                    mochi_expiring_profile_key_credential_get_expiration_time($0, contents)
                }
            }
        }
        return Date(timeIntervalSince1970: TimeInterval(timestampInSeconds))
    }
}
